File: src/services/cliProxy.ts
import { invokeTauriOrNull } from "./tauriInvoke";
import type { CliKey } from "./providers";

export type CliProxyStatus = {
  cli_key: CliKey;
  enabled: boolean;
  base_origin: string | null;
};

export type CliProxyResult = {
  trace_id: string;
  cli_key: CliKey;
  enabled: boolean;
  ok: boolean;
  error_code: string | null;
  message: string;
  base_origin: string | null;
};

export async function cliProxyStatusAll() {
  return invokeTauriOrNull<CliProxyStatus[]>("cli_proxy_status_all");
}

export async function cliProxySetEnabled(input: { cli_key: CliKey; enabled: boolean }) {
  return invokeTauriOrNull<CliProxyResult>("cli_proxy_set_enabled", {
    cliKey: input.cli_key,
    enabled: input.enabled,
  });
}

export async function cliProxySyncEnabled(base_origin: string) {
  return invokeTauriOrNull<CliProxyResult[]>("cli_proxy_sync_enabled", {
    baseOrigin: base_origin,
  });
}

File: src-tauri/src/commands/cli_proxy.rs
//! Usage: CLI proxy configuration related Tauri commands.

use crate::app_state::{ensure_db_ready, DbInitState, GatewayState};
use crate::{blocking, cli_proxy, settings};
use tauri::Emitter;
use tauri::Manager;

#[tauri::command]
pub(crate) async fn cli_proxy_status_all(
    app: tauri::AppHandle,
) -> Result<Vec<cli_proxy::CliProxyStatus>, String> {
    blocking::run("cli_proxy_status_all", move || cli_proxy::status_all(&app)).await
}

#[tauri::command]
pub(crate) async fn cli_proxy_set_enabled(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    cli_key: String,
    enabled: bool,
) -> Result<cli_proxy::CliProxyResult, String> {
    let base_origin = if enabled {
        ensure_db_ready(app.clone(), db_state.inner()).await?;

        blocking::run("cli_proxy_set_enabled_ensure_gateway", {
            let app = app.clone();
            move || {
                let state = app.state::<GatewayState>();
                let mut manager = state.0.lock().map_err(|_| "gateway state poisoned")?;
                let status = if manager.status().running {
                    manager.status()
                } else {
                    let settings = settings::read(&app).unwrap_or_default();
                    let status = manager.start(&app, Some(settings.preferred_port))?;
                    let _ = app.emit("gateway:status", status.clone());
                    status
                };

                Ok(status.base_url.unwrap_or_else(|| {
                    format!(
                        "http://127.0.0.1:{}",
                        status.port.unwrap_or(settings::DEFAULT_GATEWAY_PORT)
                    )
                }))
            }
        })
        .await?
    } else {
        blocking::run("cli_proxy_set_enabled_read_settings", {
            let app = app.clone();
            move || {
                let settings = settings::read(&app).unwrap_or_default();
                Ok(format!("http://127.0.0.1:{}", settings.preferred_port))
            }
        })
        .await?
    };

    blocking::run("cli_proxy_set_enabled_apply", move || {
        cli_proxy::set_enabled(&app, &cli_key, enabled, &base_origin)
    })
    .await
}

#[tauri::command]
pub(crate) async fn cli_proxy_sync_enabled(
    app: tauri::AppHandle,
    base_origin: String,
) -> Result<Vec<cli_proxy::CliProxyResult>, String> {
    blocking::run("cli_proxy_sync_enabled", move || {
        cli_proxy::sync_enabled(&app, &base_origin)
    })
    .await
}
